packages/cdktf-cli/lib/project-execution.ts

```diff
@@ -0,0 +1,448 @@
+    const tfClient = new TerraformCloud(
+      stack,
+      parsedStack.terraform?.backend?.remote,
+      isSpeculative,
+      sendLog
+    );
+    if (await tfClient.isRemoteWorkspace()) {
+      return tfClient;
+    }
+  }
+  return new TerraformCli(stack, sendLog);
+}
+
+const services = {
+  synthProject: async (context: ProjectContext) => {
+    const stacks = await SynthStack.synth(
+      context.synthCommand,
+      context.outDir,
+      context.workingDirectory
+    );
+
+    printAnnotations(stacks);
+
+    return stacks;
+  },
+  diff: async (context: ProjectContext) => {
+    const stack = getStack(context);
+    context.onProgress({
+      type: "STACK_SELECTED",
+      stackName: stack.name,
+    });
+    const terraform = await getTerraformClient(
+      stack,
+      true,
+      getLogCallbackForStack(context, "plan")
+    );
+    await terraform.init();
+    return terraform.plan(context.targetAction === "destroy");
+  },
+  deploy: async (context: ProjectContext) => {
+    const planFile = context.targetStackPlan?.planFile;
+    if (!planFile) {
+      throw Errors.Internal("No plan file found, diff needs to be run first");
+    }
+
+    const stack = getStack(context);
+    const terraform = await getTerraformClient(
+      stack,
+      false,
+      getLogCallbackForStack(context, "apply")
+    );
+
+    await terraform.deploy(planFile, (chunk: Buffer) => {
+      const stdout = chunk.toString("utf8");
+
+      context.onProgress({
+        type: "RESOURCE_UPDATE",
+        stackName: stack.name,
+        stateName: "deploy",
+        stdout,
+        updatedResources: parseOutput(stdout),
+      });
+    });
+  },
+  destroy: async (context: ProjectContext) => {
+    const planFile = context.targetStackPlan?.planFile;
+    if (!planFile) {
+      throw Errors.Internal("No plan file found, diff needs to be run first");
+    }
+
+    const stack = getStack(context);
+    const terraform = await getTerraformClient(
+      stack,
+      false,
+      getLogCallbackForStack(context, "destroy")
+    );
+
+    await terraform.destroy((chunk: Buffer) => {
+      const stdout = chunk.toString("utf8");
+
+      context.onProgress({
+        type: "RESOURCE_UPDATE",
+        stackName: stack.name,
+        stateName: "deploy",
+        stdout,
+        updatedResources: parseOutput(stdout),
+      });
+    });
+  },
+  gatherOutput: async (context: ProjectContext) => {
+    if (context.targetAction === "destroy") {
+      return Promise.resolve({});
+    }
+
+    const stack = getStack(context);
+    const terraform = await getTerraformClient(
+      stack,
+      false,
+      getLogCallbackForStack(context, "output")
+    );
+    const outputs = await terraform.output();
+    const outputsByConstructId = getConstructIdsForOutputs(
+      JSON.parse(stack.content),
+      outputs
+    );
+
+    return Promise.resolve({ outputs, outputsByConstructId });
+  },
+};
+
+const guards = {
+  onTargetAction: (
```

Comment:
I had a hard time grasping what this does (when reading the state machine config). Could this be improved somehow to create a more speaking name when using the guard?
Maybe something like `targetActionIs`?
That way the usage could read (with reordered lines):
```
cond: {
  name: "targetActionIsSynth",
  type: "targetActionIs",
  value: "synth",
}
```